Limit the size of the result window to a dynamic property

nik9000 opened the pull request:
Requesting a million hits, or page 100,000 is always a bad idea, but users may not be aware of this. This adds a per-index limit on the maximum `size + from` that can be requested which defaults to 10,000.

This should not interfere with deep-scrolling.

Closes #9311

nik9000 commented:
Still a bit of a work in progress but ok for review.

nik9000 commented:
> Still a bit of a work in progress but ok for review.

And now it seems to be working pretty well.

dakrone commented on core/src/main/java/org/elasticsearch/search/internal/DefaultSearchContext.java:
```diff
@@ -179,12 +193,20 @@ public void doClose() {
+            long resultWindow = from + size;
+            // We need settingsService's view of the settings because its dynamic.
+            // indexService's isn't.
+            int maxResultWindow = indexService.settingsService().getSettings().getAsInt(MAX_RESULT_WINDOW, Defaults.MAX_RESULT_WINDOW);
```
So I know we don't expect anyone to set this to a crazy high value, but I think we should make it a `long` **just** because someone is liable to set it to `Integer.MAX_VALUE + 1` and then wonder why it shows negative due to wrapping.

jpountz replied:
+1

nik9000 replied:
I configure the settings validator as POSITIVE_INTEGER below. Is that enough?

clintongormley commented:
@nik9000 could you add some description of the PR to make it easier to understand? This PR is what the change log links to.

clintongormley commented on core/src/main/java/org/elasticsearch/search/internal/DefaultSearchContext.java:
```diff
@@ -179,12 +193,20 @@ public void doClose() {
+                                + "This limit can be set by changing the [" + DefaultSearchContext.MAX_RESULT_WINDOW
+                                + "] index level parameter.");
             }
         }
```
Nice!

clintongormley commented:
Could you document this setting here as well please: https://www.elastic.co/guide/en/elasticsearch/reference/2.0/index-modules.html

jpountz commented on docs/reference/search/request/from-size.asciidoc:
```diff
@@ -19,3 +19,7 @@ defaults to `10`.
 --------------------------------------------------
+
+Note that `from` + `size` can not be more than the `index.max_result_window`
+index setting which defaults to 10,000. See the scroll api for more efficient
```
maybe put a link from "scroll" to the scroll docs?

jpountz commented:
Woohoo! LGTM

nik9000 commented:
> @nik9000 could you add some description of the PR to make it easier to understand? This PR is what the change log links to.

Done.

nik9000 commented:
> Could you document this setting here as well please: https://www.elastic.co/guide/en/elasticsearch/reference/2.0/index-modules.html

Done

nik9000 commented:
Ok - I think this is pretty much ready. There is still an open question above about long for the setting. I use a POSITIVE_INTEGER validator so I think that is enough, right? I can add a test to the setting that tries to set it to Integer.MAX_VALUE + 1 and notices the error if you'd like.

nik9000 commented:
Ping @dakrone and @jpountz - this has one open discussion point but seems otherwise ready. Can you have a look above? Its the long vs int discussion.

nik9000 commented:
Ping @dakrone and @jpountz again - I think this is just waiting on an answer from you to the question above about int vs long?

jpountz commented:
Sorry I totally missed this message. The validator should work indeed, +1 to push.

nik9000 commented:
> Sorry I totally missed this message. The validator should work indeed, +1 to push.

Cool! I've squashed, rebased, and made some minor corrections to deal with COUNT not being a search type any more.

nik9000 commented:
Pushed to master and 2.1.

clintongormley commented:
w00t
